### PR TITLE
Add host firewall

### PR DESCRIPTION
**Pull request by richbayliss**

### CODE REVIEW

File: src/lib/firewall.ts [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+
+import * as dbFormat from '../device-state/db-format';
+
+export const initialised = (async () => {
+	await config.initialized;
+	await applyFirewall();
+
+	// apply firewall whenever relevant config changes occur...
+	config.on('change', async ({ firewallMode, localMode }) => {
+		if (firewallMode || localMode !== undefined) {
```
Comment (CameronDiver) [inline src/lib/firewall.ts:16]:
```suggestion
		if (firewallMode || localMode != null) {
```
I always encourage this until everything in TS in the supervisor (and even then I'd probably continue to).

File: src/lib/firewall.ts [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+			applyFirewall({ firewallMode, localMode });
+		}
+	});
+})();
+
+const BALENA_FIREWALL_CHAIN = 'BALENA-FIREWALL';
+const LOG_PREFIX = '\u{1F525}';
```
Comment (CameronDiver) [inline src/lib/firewall.ts:23]: Is this the fire emoji? You can just add it directly

File: src/lib/firewall.ts [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+];
+
+const standardPolicy: iptables.Rule[] = [
+	{
+		comment: 'Locally-sourced traffic',
+		action: '-I',
+		matches: ['-m addrtype', '--src-type LOCAL'],
+		target: 'ACCEPT',
+	},
+	{
```
Comment (CameronDiver) [inline src/lib/firewall.ts:70]: What's this rule for?

File: src/lib/firewall.ts [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+			supervisorAccessRules.push({
+				comment: 'Supervisor API',
+				action: '-A',
+				proto: 'tcp',
+				matches: [`--dport ${port}`, `-i ${intf}`],
+				target: 'ACCEPT',
+				applyIfExists: false,
+			});
+		});
+	}
```
Comment (CameronDiver) [inline src/lib/firewall.ts:111]:
```suggestion

			supervisorAccessRules.push({
				comment: 'Supervisor API',
				action: '-A',
				proto: 'tcp',
				matches: [`--dport ${port}`].concat(localMode? [] : [`-i ${intf}`']),
				target: 'ACCEPT',
				applyIfExists: false,
			});
	}
```

File: src/lib/firewall.ts [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+		'listenPort',
+		'firewallMode',
+		'localMode',
+	]);
+
+	// populate missing config elements...
+	const { listenPort, firewallMode, localMode } = _.defaults(
+		opts,
+		currentConfig,
+	);
```
Comment (CameronDiver) [inline src/lib/firewall.ts:136]:
```suggestion
	const { listenPort, firewallMode, localMode } = {
		...opts,
		...currentConfig,
	};
```

File: src/lib/firewall.ts [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+		localMode,
+		constants.allowedInterfaces,
+		listenPort,
+	);
+
+	// apply the firewall rules...
+	await exports.applyFirewallMode(firewallMode ?? '');
+}
+
+export const AllowedModes = ['on', 'off', 'auto'];
```
Comment (CameronDiver) [inline src/lib/firewall.ts:149]: Either `ALLOWED_MODES` or `allowedModes`, only type names should be capitalized and camel case.

File: src/lib/iptables.ts [hunk 7]
```diff
@@ -1,66 +1,170 @@
...
+	if (rule.comment) {
+		args.push('-m comment');
+		args.push(`--comment "${rule.comment}"`);
+	}
+	if (rule.target) {
+		args.push(`-j ${rule.target}`);
+	}
+
+	if (!rule.family) {
+		rule.family = 4;
```
Comment (CameronDiver) [inline src/lib/iptables.ts:62]: Weren't we going to apply to both by default?

File: src/lib/iptables.ts [hunk 8]
```diff
@@ -1,66 +1,170 @@
...
+				...rule,
+				...{
+					family: 4,
+				},
+			},
+			adaptor,
 		);
+	}
+
+	if (rule.applyIfExists !== undefined) {
```
Comment (CameronDiver) [inline src/lib/iptables.ts:132]:
```suggestion
	if (rule.applyIfExists != null) {
```

File: src/lib/iptables.ts [hunk 9]
```diff
@@ -1,66 +1,170 @@
...
-		),
-	)
-		.then(() =>
-			clearAndAppendIptablesRule(
-				`OUTPUT -p tcp --sport ${port} -m state --state ESTABLISHED -j ACCEPT`,
-			),
-		)
-		.then(() =>
-			clearAndAppendIptablesRule(`INPUT -p tcp --dport ${port} -j REJECT`),
+async function applyRule(rule: Rule, adaptor: RuleAdaptor) {
```
Comment (CameronDiver) [inline src/lib/iptables.ts:118]: Can you either split this function or add some comments please? Following what's going on here isn't too easy.

File: src/lib/firewall.ts (hunk 3 above)
Comment (richbayliss) [reply src/lib/firewall.ts:70]: Locally-sourced traffic -- traffic coming into the device from itself; stuff like DNS lookups that use 127.0.0.2

File: src/lib/firewall.ts (hunk 4 above)
Comment (richbayliss) [reply src/lib/firewall.ts:111]: there's also a difference in that the non-localmode path has to create this rule for each interface

Comment (richbayliss) [reply src/lib/firewall.ts:111]: also, you cannot pass them in one hit -- `multiple -i flags not allowed`

File: src/lib/iptables.ts (hunk 7 above)
Comment (richbayliss) [reply src/lib/iptables.ts:62]: We do, just not here -- this is the adaptor which applies the rule and it _needs_ a family is one is missing. However, we don't call this externally, we have another function which wraps/sanitises the *Rule* object and it handles the extra apply if it's missing a family 👍

File: src/lib/iptables.ts [hunk 10]
```diff
@@ -1,66 +1,184 @@
...
+import { child_process } from 'mz';
 
-function clearIptablesRule(rule: string): Bluebird<void> {
-	return applyIptablesArgs(`-D ${rule}`);
+export class IPTablesRuleError extends Error {}
+
+export interface Rule {
+	id?: number;
+	family?: 4 | 6;
+	action?: '-L' | '-I' | '-A' | '-D' | '-C' | '-N' | '-F' | '-X';
```
Comment (wrboyce) [inline src/lib/iptables.ts:9]: I'd be tempted to use an Enum here

```typescript
enum RuleAction {
    List = '-L',
    Insert = '-I',
    // etc
}
```

**Pull request merged**